### PR TITLE
graph/encoding/graphql: add new package for decoding GraphQL output

### LINKED ISSUE #72: graph/encoding: add support for parsing JSON-represented graphs
GraphQL outputs JSON in response to queries. We should be able to read these results in to allow operation on those graphs.

This is one of the use cases prompting the change from `int` to `int64` IDs.

There are two cases for the GraphQL output as far as we are concerned, with uid in which case we respect the provide value, and without when we need to make it up.

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

Depends on #168.
Closes #72.

### CODE REVIEW

File: graph/encoding/graphql/decode.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	SetLabel(string)
+}
+
+type generator struct {
+	dst encoding.Builder
+
+	// uidName is the name of the UID field in the source JSON.
+	uidName string
+
+	nodes map[string]graph.Node
```
Comment (mewmew) [inline graph/encoding/graphql/decode.go]: Add a comment stating the key of the map; e.g.

```go
// nodes maps from node value to graph node.
```

Is this comment correct? Are node values unique?

File: graph/encoding/graphql/decode.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+				}
+				return fmt.Errorf("graphql: no UID for node: `%s`", &buf)
+			}
+		} else {
+			var v interface{}
+			err = json.Unmarshal(next, &v)
+			if err != nil {
+				return err
+			}
+			value := fmt.Sprint(v)
```
Comment (mewmew) [inline graph/encoding/graphql/decode.go]: What does the call to `fmt.Sprint` do? Can it be replaced with a type assertion for `fmt.Stringer`?

File: graph/encoding/graphql/decode.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		}
+
+	default:
+		var v interface{}
+		err := json.Unmarshal(src, &v)
+		if err != nil {
+			return err
+		}
+		if attr == g.uidName {
+			value := fmt.Sprint(v)
```
Comment (mewmew) [inline graph/encoding/graphql/decode.go]: Same as comment above.

File: graph/encoding/graphql/decode.go (hunk 1 above)
Comment (kortschak) [inline graph/encoding/graphql/decode.go]: Done.

UIDs are unique in the input text.

File: graph/encoding/graphql/decode.go (hunk 2 above)
Comment (kortschak) [inline graph/encoding/graphql/decode.go]: Everything can be represented as a string, `fmt.Sprint(v)` does this conversion. The two things we are likely to get here are a `string` or an `int`. Neither of these implement `fmt.Stringer`.

File: graph/encoding/graphql/decode.go (hunk 3 above)
Comment (kortschak) [inline graph/encoding/graphql/decode.go]: Same answer.

**Comment by kortschak** [comment]
PTAL

File: graph/encoding/graphql/decode.go (hunk 2 above)
Comment (mewmew) [inline graph/encoding/graphql/decode.go]: Ok.

**Comment by mewmew** [comment]
LGTM

**Comment by kortschak** [comment]
Can you re-approve please.

**Comment by mewmew** [comment]
I tried to have a look around, but can't find the option to do so in the web UI.

Where to re-approve?

**Comment by mewmew** [comment]
It says

>  kortschak dismissed mewmew’s stale review via 1b863d0 7 hours ago

and now, the review button is gone for what it looks like.

**Comment by mewmew** [comment]
I think the reason may be the force push,

> We went looking everywhere, but couldn’t find those commits.

**Comment by kortschak** [comment]
You need to be in the [files changed tab](https://github.com/gonum/gonum/pull/169/files), where you'll see green "Review changes" button/menu/dialogue. The review is dismissed by any additional push.

**Comment by mewmew** [comment]
> You need to be in the files changed tab, where you'll see green "Review changes" button/menu/dialogue. The review is dismissed by any additional push.

Ok, I'll take a look.

**Comment by mewmew** [comment]
For future notes,

@kortschak Did you do a force push, or how did you push the latest incremental commit?

The benefit of uploading incremental commits, is that it is easier to review the changes, then you see the exact lines which change. Also, when the PR is complete, you may hit [squash and merge](https://github.com/blog/2141-squash-your-commits) to turn those several commits into a single one.

**Comment by kortschak** [comment]
It was forced. In this PR there was very minimal change (the changes are visible from the GUI - it's unfortunate githubs git GC is so aggressive in removing those commits - it did not use to be) in addressing the comments - and significantly, the PR would not pass unless rebased onto the changes in #168 merged into master. Normally I do do incremental and retain the history (at least since the addition of the squash option on the GUI).

**Comment by mewmew** [comment]
I see. Thanks for the background.

**Pull request merged**